archive/import.go

```diff
@@ -2,156 +2,350 @@ package archive
 
 import (
 	"context"
-	"fmt"
 	"io/ioutil"
 	"os"
 
-	"github.com/brimsec/zq/driver"
+	"github.com/brimsec/zq/expr"
 	"github.com/brimsec/zq/microindex"
 	"github.com/brimsec/zq/pkg/bufwriter"
 	"github.com/brimsec/zq/pkg/fs"
 	"github.com/brimsec/zq/pkg/iosrc"
 	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/proc/sort"
+	"github.com/brimsec/zq/proc/spill"
 	"github.com/brimsec/zq/zbuf"
 	"github.com/brimsec/zq/zio/zngio"
 	"github.com/brimsec/zq/zng"
 	"github.com/brimsec/zq/zng/resolver"
-	"github.com/brimsec/zq/zqd/api"
-	"github.com/brimsec/zq/zql"
+	"go.uber.org/multierr"
 )
 
+// ImportBufSize specifies the max size of the records buffered during import
+// before they are flushed to disk.
+var ImportBufSize = int64(sort.MemMaxBytes)
+
 // The below are vars for unit testing.
 var (
 	importLZ4BlockSize     = zngio.DefaultLZ4BlockSize
 	importStreamRecordsMax = zngio.DefaultStreamRecordsMax
 )
 
-type importDriver struct {
-	ark             *Archive
-	ctx             context.Context
-	dataFile        dataFile
-	dataFileWriter  *zngio.Writer
-	firstTs         nano.Ts
-	indexBuilder    *zng.Builder
-	indexTempPath   string
-	indexTempWriter *zngio.Writer
-	lastTs          nano.Ts
-	needIndexWrite  bool
-	rcount          int
-	tsDir           iosrc.URI
-	zctx            *resolver.Context
+func Import(ctx context.Context, ark *Archive, zctx *resolver.Context, r zbuf.Reader) error {
+	w := newImportWriter(ctx, ark)
+	err := zbuf.CopyWithContext(ctx, w, r)
+	if closeErr := w.close(); err == nil {
+		err = closeErr
+	}
+	return err
 }
 
-func (d *importDriver) newWriter(rec *zng.Record) error {
-	d.firstTs = rec.Ts()
-	d.rcount = 0
+// importWriter is a zbuf.Writer the partitions records by day into the
+// appropriate tsDirWriter. importWriter keeps track of the overall memory
+// footprint of the collection of tsDirWriter and instructs the tsDirWriter
+// with the largest footprint to spill its records to a temporary file on disk.
+//
+// XXX importWriter does not currently keep track of size of records written
```

Comment:
How about add a note saying there's no time based flushing yet, either?

Reply:
Should I file an issue for both of these?

Reply:
Issues created.